sql/analyzer/resolve_unions.go

```diff
@@ -22,66 +22,83 @@ import (
 	"github.com/dolthub/go-mysql-server/sql"
 	"github.com/dolthub/go-mysql-server/sql/expression"
 	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/transform"
 )
 
 // resolveUnions resolves the left and right side of a union node in isolation.
-func resolveUnions(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, error) {
+func resolveUnions(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, transform.TreeIdentity, error) {
 	if n.Resolved() {
-		return n, nil
+		return n, transform.SameTree, nil
 	}
 	// Procedures explicitly handle unions
 	if _, ok := n.(*plan.CreateProcedure); ok {
-		return n, nil
+		return n, transform.SameTree, nil
 	}
 
-	return plan.TransformUp(n, func(n sql.Node) (sql.Node, error) {
-		switch n := n.(type) {
+	return transform.Node(n, func(node sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		switch n := node.(type) {
 		case *plan.Union:
 			subqueryCtx, cancelFunc := ctx.NewSubContext()
 			defer cancelFunc()
 
-			left, err := a.analyzeThroughBatch(subqueryCtx, n.Left(), scope, "default-rules")
+			left, sameL, err := a.analyzeThroughBatch(subqueryCtx, n.Left(), scope, "default-rules")
 			if err != nil {
-				return nil, err
+				return nil, transform.SameTree, err
 			}
 
-			right, err := a.analyzeThroughBatch(subqueryCtx, n.Right(), scope, "default-rules")
+			right, sameR, err := a.analyzeThroughBatch(subqueryCtx, n.Right(), scope, "default-rules")
 			if err != nil {
-				return nil, err
+				return nil, transform.SameTree, err
 			}
 
-			return n.WithChildren(StripPassthroughNodes(left), StripPassthroughNodes(right))
+			if sameL && sameR {
```

Comment:
Should just ignore the `same` return params entirely. This function should always run if the resolved check above passes.

Reply:
oh that's clever, might be useful for knocking out other false positive rules, where we return transform.NewTree when nothing changed